### ISSUE TITLE
Feedback on Notebook Extension API

### ISSUE THREAD

**Issue by darknoon** [microsoft/vscode#98257]
I have been working with the proposed Notebook extension API to provide a javascript notebook in the style of observable notebooks (for prototyping, I am using their libraries). I don't have a ton of experience with VSCode extension development, so maybe I have missed something obvious.

<img width="933" alt="Screen Shot 2020-05-20 at 12 41 00 PM" src="https://user-images.githubusercontent.com/355540/82473191-351c2e80-9a97-11ea-91fa-879d6ed7910a.png">

### Capabilities

- It was easy to get started with the content provider API and the loading API makes sense.

- There doesn't seem to be a way to track changes to notebook cells before they are executed (is this meant to be handled by language extensions instead?). For example, it would be cool to provide speculative execution à la Chrome's console.

- Since an Observable notebook will continue executing and returning results, maybe should have an API to choose whether document is dirtied by continued output (ticking up counters etc)? To work around this (and allow HTML output), I'm switching everything over to a custom output renderer.

- I like that I can define whether the cell is executing (showing rotating arrows), because in observable notebooks other cells can execute in response to changing another cell

- How do I attach DevTools to the Webview that is hosting the cell outputs? It was annoying to be flying blind as to what is happening with outputs.

- Notebook open/close events weren't being called. I think this [might be fixed](https://github.com/microsoft/vscode/commit/cc151dd103648a309754fac148bd182395973d1c)

- Have been using `Map<Cell, MyObject>` to do internal bookkeeping, is this safe in general? I noticed that the Julia code [is using Uri.toString()](https://github.com/julia-vscode/julia-vscode/blob/efcc643633016c7991774dcce5fc4d055041d067/src/)

- Can this work with remote development? Any particular things I need to ensure that it will be safe?

#### Syncing outputs with webviews

To render html / not cause the document to be dirtied every time an output changed, I switched to a custom `vscode.NotebookOutputRenderer` with a defined MIME type.

- I was stuck with a nasty race condition between the code that shows the output from the cell and needs to observe events from the kernel

  1. Document starts executing
  1. "Kernel" starts (currently this is still part of the content provider)
  1. cells start going from no outputs to either an custom output, an async "executing" or a native error output
  1. Output webview starts
  1. Output provider starts, provides its preloads (script that handles observation)
  1. Output provider is called for some of the cells (order can vary) to provide an HTML string
  1. HTML string is added to document, js executes
  1. At this point, there is a race between loading the observation script `<script src="${resourceUri.toString()}" />` that provides the observeCell api, and `<script>observeCell("my-id-123")</script>` tag with within the HTML string. For some of the cells, the script gets called before the script is loaded. For others it is after. I suspect that the [`domEval` code is to blame](https://github.com/microsoft/vscode/blob/2d6df55ad14d14922c8989d88517a3d8bba0023f/src/vs/workbench/contrib/notebook/browser/view/renderers/backLayerWebView.ts#L236). I'm curious as to why this is here.
  1. The preloads also can load **after** some of the cells run their script tag js, so hard to ensure that my library is available. Should I be using ES6 import to load the Uri referencing the extension resource?

So:

- Cell outputs are ready, but output webview HTML don't exist yet. I can embed the initial value into the cell output, or queue the events for once the cell "boots" and send a "boot" message from the cell. The issue with the "boot" message is that I need to make sure I have subscribed with `activeNotebookEditor.onDidReceiveMessage()` before the "boot" messages get sent, but I don't have a handle to `activeNotebookEditor` yet.

- It's hard to know from the kernel / content provider when the output webview is ready so we can start sending the cells value updated events and getting boot events back. I made a [nasty hack](https://github.com/darknoon/fluid-notebook/blob/d594e91e450efb18c686fe720c88871131cd31e5/src/hackNotebookEditorEvent.ts) that checks every 500ms for the `vscode.notebook.activeNotebookEditor` to change to one that represents the current document.

- As a workaround, I just search for any "unbooted" cells in the cell script, but I think it's still unrelible.

Please let me know if I should be doing this differently!

### UI

- I miss some of the "clean" look of observable notebooks. I think the overall UI chrome could be dialed back a bit like collaboratory or a regular text editor view.

- Language selector takes up a lot of space

### Bugs

- Sometimes the notebook is slow to load / gets stuck at a blank white screen. Can't get it to reproduce reliably though.

- If a custom HTML output changes frequently (my notebook was changing it every second), VSCode locks up and I had to force-quit it.

### Naming / nits

- `CellStreamOutput` seems to be a misnomer since it just provides text output (that you can modify of course)

I'm happy to break these out into separate issues if that would be more helpful. Thanks!

- VSCode Version: 1.46.0-insider
- OS Version: macOS 10.15.4 (19E287)

Perhaps relevant developers:

@rebornix @mjbvz @roblourens

**Comment by rebornix** [comment]
@darknoon thanks for your insightful feedback, good work! We will create issues for the bugs / ux feedback later on, now I can answer a few questions posted above to unblock you from continuing your exploration with our API 

---

> There doesn't seem to be a way to track changes to notebook cells before they are executed (is this meant to be handled by language extensions instead?). For example, it would be cool to provide speculative execution à la Chrome's console.

Currently cells in a notebook document will be opened as normal text editor once it's rendered or its content is modified. That means you can listen to `vscode.workspace.onDidOpenTextDocument`, etc to access its content.

> Since an Observable notebook will continue executing and returning results, maybe should have an API to choose whether document is dirtied by continued output

Yes it's part of our plan to allow extensions to control if outputs can make the editor dirty.

> How do I attach DevTools to the Webview that is hosting the cell outputs? It was annoying to be flying blind as to what is happening with outputs.

F1 -> Developer: Open Webview Developer Tools

> Notebook open/close events weren't being called

Right, it's fixed in latest Insiders

> Have been using Map<Cell, MyObject> to do internal bookkeeping, is this safe in general? I noticed that the Julia code is using Uri.toString()

`NotebookCell.uri` can be used as the unique ID for a cell during one session (before the editor is closed completely or the window is reloaded).

> Can this work with remote development

Yes, almost. We currently don't support running custom renderer as a UI extension and running the notebook content provider in remote. It's a bug since we are now handling output transformation in ext host. The fix will come next week.

> Syncing outputs with webviews

Good questions and I can understand the pain you went through to make it work properly. Right now we didn't introduce any life cycle concept yet to the renderers (@connor4312 has good proposals for it https://github.com/microsoft/vscode/issues/95361#issuecomment-630925871), thus the renderer needs to solve the racing between preload scripts loading and output evaluation.

THe catch here is the scripts loading are handled by the browser, and we (vscode core) don't know if the initialization of the preload scripts are done or not. The pattern we use right now is 

```
# preload scripts
(function () {
    ...
    initialization() ...
    window.observeCell = () => { ... }
    // search for all cells are not rendered yet
    cells.forEach(cell => observeCell(cell))
})();

# output cell

if (window.observeCell) { window.observeCell(cell)}
```

**Comment by darknoon** [comment]
Thank you @rebornix! I worked on adopting the new `Kernel` api and `visibleEditorsChanged()` yesterday, which help a lot.

### import

This does resolve the race condition:

```html
<script type="module">
import { subscribeCell } from "vscode-webview-resource://<TOKEN>/blah/out/webview/esm/webview.js"
</script>
```

I had to make a separate `tsconfig.json` so I could build that directory with `    "module": "ESNext",` which isn't ideal but is better than the earlier workarounds. It would be nice if I could specify this resource as a preload (but preloads doesn't allow you to  `type="module"` so it wouldn't load). I haven't tried `async import(…)` yet, which supposedly works inside of non-module scripts.

"scripts loading are handled by the browser" -> true, but if the scripts hadn't been hoisted by the VSCode `domEval`, they would have loaded in the normal HTML document order and this wouldn't have been an issue.

----

### `NotebookCell.uri`
 I'll try to use this instead of creating my own ident and report back!

### Document → Editor

It feels slightly awkward to ask `vscode.notebook.visibleNotebookEditors.find(…)` in the `render()` method… if I don't have the `NotebookEditor` available at that point, I won't be able to create the `asWebviewUri`, so could that be passed in as a parameter?

Or perhaps the current editor for a document could be exposed on `document` as an optional parameter, making it easier to subscribe to "started editing" or whatever you call that transition?

Relevant code:

```typescript
  render(
    notebook: vscode.NotebookDocument,
    output: NotebookOutput,
    mimeType: string
  ): string {
    if (mimeType !== OutputMimeType) {
      throw new Error(`Asked to render unknown MIME type ${mimeType}`);
    }
    const data = output.data[OutputMimeType];
    const { ident } = data as { ident: string };

    const editor = vscode.notebook.visibleNotebookEditors.find(
      (ne) => ne.document === notebook
    );
    const uri = editor?.asWebviewUri(this.localResourceUri("esm/webview.js"));

    if (uri === undefined) {
      console.error(
        "Need to output HTML for cell, but we don't have an editor!"
      );
    }

    return /* html */ `
        <div class="darknoon-fluid-notebook-output unbooted" id="${ident}">Unbooted <code>${ident}</code></div>
        <script type="module">
          ${uri ? `import {subscribeCell} from "${uri?.toString()}";` : ""}
          if (typeof subscribeCell !== "undefined") {
            subscribeCell("${ident}");
          } else {
            console.error("Couldn't subscribe for id ${ident}. Waiting for our preload script.");
          }
        </script>
      `;
  }
}
```

**Comment by rebornix** [comment]
> It feels slightly awkward to ask vscode.notebook.visibleNotebookEditors.find(…) in the render() method… if I don't have the NotebookEditor available at that point, I won't be able to create the asWebviewUri, so could that be passed in as a parameter?


@darknoon good catch, the render API was designed before we introduced the messaging channel on the `NotebookEditor`. It's better to pass in `NotebookEditor` other than `NotebookDocument` though I'm not sure if that would be the final decision. We might go with passing in a comm object which allows you to convert Uri and communicate with the webview in the editor, other than exposing the message channel to every one in the extension host.

**Comment by jrieken** [comment]
This issue has become stale, closing in favour of various other API discussions that we are having:
